pkg/server/list_users_test.go

```diff
@@ -0,0 +1,217 @@
+		},
+		{
+			name: "invalid_relation",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "owner", // invalid relation
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model:             model,
+			expectedErrorCode: codes.Code(2022),
+		},
+		{
+			name: "contextual_tuple_invalid_object_type",
+			req: &openfgav1.ListUsersRequest{
+				Object:      &openfgav1.Object{Type: "document", Id: "1"},
+				Relation:    "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{{Type: "user"}},
+				ContextualTuples: &openfgav1.ContextualTupleKeys{
+					TupleKeys: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("invalid_object_type:1", "viewer", "user:will"),
+					},
+				},
+			},
+			model:             model,
+			expectedErrorCode: codes.Code(2027),
+		},
+		{
+			name: "contextual_tuple_invalid_user_type",
+			req: &openfgav1.ListUsersRequest{
+				Object:      &openfgav1.Object{Type: "document", Id: "1"},
+				Relation:    "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{{Type: "user"}},
+				ContextualTuples: &openfgav1.ContextualTupleKeys{
+					TupleKeys: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("document:1", "viewer", "invalid_user_type:will"),
+					},
+				},
+			},
+			model:             model,
+			expectedErrorCode: codes.Code(2027),
+		},
+		{
+			name: "contextual_tuple_invalid_relation",
+			req: &openfgav1.ListUsersRequest{
+				Object:      &openfgav1.Object{Type: "document", Id: "1"},
+				Relation:    "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{{Type: "user"}},
+				ContextualTuples: &openfgav1.ContextualTupleKeys{
+					TupleKeys: []*openfgav1.TupleKey{
+						tuple.NewTupleKey("document:1", "invalid_relation", "user:will"),
+					},
+				},
+			},
+			model:             model,
+			expectedErrorCode: codes.Code(2027),
+		},
+	}
+
+	storeID := ulid.Make().String()
+	for _, test := range tests {
+		ds := memory.New()
+		t.Cleanup(ds.Close)
+		model := testutils.MustTransformDSLToProtoWithID(test.model)
+
+		t.Run(test.name, func(t *testing.T) {
+			typesys, err := typesystem.NewAndValidate(context.Background(), model)
+			require.NoError(t, err)
+
+			err = ds.WriteAuthorizationModel(context.Background(), storeID, model)
+			require.NoError(t, err)
+
+			s := MustNewServerWithOpts(
+				WithDatastore(ds),
+			)
+			s.experimentals = []ExperimentalFeatureFlag{ExperimentalEnableListUsers}
+			t.Cleanup(s.Close)
+
+			ctx := typesystem.ContextWithTypesystem(context.Background(), typesys)
+
+			test.req.AuthorizationModelId = model.GetId()
+			test.req.StoreId = storeID
+
+			_, err = s.ListUsers(ctx, test.req)
+			e, ok := status.FromError(err)
+			require.True(t, ok)
+
+			fmt.Println("Actual", e.Code().String(), "Expected", test.expectedErrorCode.String())
+			require.Equal(t, test.expectedErrorCode, e.Code())
+		})
+	}
+}
+func TestExperimentalListUsers(t *testing.T) {
```

Comment:
Can we add a test where if you pass an explicit model ID that can't be found, you get a http 400?